remill/Arch/X86/Semantics/MMX.h

```diff
@@ -73,6 +149,784 @@ DEF_SEM(PUNPCKHQDQ, D dst, S1 src1, S2 src2) {
+    for(size_t i = 0; i < vec_count; i++){
+        auto sum_elem = UAdd(UInt32(UExtractV16(lhs_vec, i)), UInt32(UExtractV16(rhs_vec, i))) ;
+        auto sum = UAdd(sum_elem, decltype(sum_elem)(1));
+        dst_vec = UInsertV16(dst_vec, i, UInt16(UShr(sum, decltype(sum)(1))));
+    }
+	UWriteV16(dst, dst_vec);
+}
+DEF_ISEL(PAVGW_MMXq_MMXq) = PAVGW<V64W, V64, V64>;
+DEF_ISEL(PAVGW_MMXq_MEMq) = PAVGW<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PHSUBW, D dst, S1 src1, S2 src2) {
+    auto lhs_vec = SReadV16(src1);
+	auto rhs_vec = SReadV16(src2);
+	auto dst_vec = SClearV16(SReadV16(dst));
+	auto vec_count = NumVectorElems(lhs_vec);
+	for(size_t i = 0; i < vec_count; i = i+2) {
+		dst_vec = SInsertV16(dst_vec, i/2, SSub(SExtractV16(lhs_vec, i), SExtractV16(lhs_vec, i+1)));
+	}
+	for(size_t i = 0; i < NumVectorElems(rhs_vec); i = i+2) {
+		dst_vec = SInsertV16(dst_vec, (i+vec_count)/2, SSub(SExtractV16(rhs_vec, i), SExtractV16(rhs_vec, i+1)));
+	}
+	SWriteV16(dst, dst_vec);
+}
+DEF_ISEL(PHSUBW_MMXq_MMXq) = PHSUBW<V64W, V64, V64>;
+DEF_ISEL(PHSUBW_MMXq_MEMq) = PHSUBW<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PHSUBD, D dst, S1 src1, S2 src2) {
+    auto lhs_vec = SReadV32(src1);
+	auto rhs_vec = SReadV32(src2);
+	auto dst_vec = SClearV32(SReadV32(dst));
+	auto vec_count = NumVectorElems(lhs_vec);
+	for(size_t i = 0; i < vec_count; i = i+2) {
+		dst_vec = SInsertV32(dst_vec, i/2, SSub(SExtractV32(lhs_vec, i), SExtractV32(lhs_vec, i+1)));
+	}
+	for(size_t i = 0; i < NumVectorElems(rhs_vec); i = i+2) {
+		dst_vec = SInsertV32(dst_vec, (i+vec_count)/2, SSub(SExtractV32(rhs_vec, i), SExtractV32(rhs_vec, i+1)));
+	}
+	SWriteV32(dst, dst_vec);
+}
+DEF_ISEL(PHSUBD_MMXq_MMXq) = PHSUBD<V64W, V64, V64>;
+DEF_ISEL(PHSUBD_MMXq_MEMq) = PHSUBD<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PMAXSW, D dst, S1 src1, S2 src2) {
+    auto lhs_vec = SReadV16(src1);
+    auto rhs_vec = SReadV16(src2);
+    auto dst_vec = SClearV16(SReadV16(dst));
+
+    // Compute MAX of words
+    auto vec_count = NumVectorElems(lhs_vec);
+    for(size_t i = 0; i < vec_count; i++) {
+        auto max = Select(SCmpGt(SExtractV16(lhs_vec, i), SExtractV16(rhs_vec, i)), SExtractV16(lhs_vec, i), SExtractV16(rhs_vec, i));
+        dst_vec = SInsertV16(dst_vec, i, max);
+    }
+    SWriteV16(dst, dst_vec);
+}
+DEF_ISEL(PMAXSW_MMXq_MMXq) = PMAXSW<V64W, V64, V64>;
+DEF_ISEL(PMAXSW_MMXq_MEMq) = PMAXSW<V64W, V64, MV64>;
+
+template <typename D, typename S1, typename S2>
+DEF_SEM(PMAXUB, D dst, S1 src1, S2 src2) {
+    auto lhs_vec = UReadV8(src1);
+    auto rhs_vec = UReadV8(src2);
+    auto dst_vec = UClearV8(UReadV8(dst));
+
+    // Compute MAX of bytes
+    auto vec_count = NumVectorElems(lhs_vec);
+    for(size_t i = 0; i < vec_count; i++) {
+        auto max = Select(UCmpGt(UExtractV8(lhs_vec, i), UExtractV8(rhs_vec, i)), UExtractV8(lhs_vec, i), UExtractV8(rhs_vec, i));
```

Comment:
You know, this is coming up sufficiently often that I'd like you to extend `Operators.h` with a `SMax`, `SMin`, `UMax`, and `UMin` and then update all the code accordingly. That will really simplify things :-D